client/handle_http.go

```diff
@@ -1120,25 +1184,124 @@ func runTransferWorker(ctx context.Context, workChan <-chan *clientTransferFile,
 	}
 }
 
+// If there are multiple potential attempts, try to see if we can quickly eliminate some of them
+//
+// Attempts a HEAD against all the endpoints simultaneously.  Put any that don't respond within
+// a second behind those that do respond.
+func sortAttempts(ctx context.Context, path string, attempts []transferAttemptDetails) (size int64, results []transferAttemptDetails) {
+	size = -1
+	if len(attempts) < 2 {
+		results = attempts
+		return
+	}
+	transport := config.GetTransport()
+	headChan := make(chan struct {
+		idx  int
+		size uint64
+		err  error
+	})
+	defer close(headChan)
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
+	defer cancel()
+	for idx, transferEndpoint := range attempts {
+		tUrl := *transferEndpoint.Url
+		tUrl.Path = path
+
+		go func(idx int, tUrl string) {
+			headClient := &http.Client{Transport: transport}
+			headRequest, _ := http.NewRequestWithContext(ctx, "HEAD", tUrl, nil)
+			var headResponse *http.Response
+			headResponse, err := headClient.Do(headRequest)
+			if err != nil {
+				headChan <- struct {
+					idx  int
+					size uint64
+					err  error
+				}{idx, 0, err}
+				return
+			}
+			headResponse.Body.Close()
+			contentLengthStr := headResponse.Header.Get("Content-Length")
+			size := int64(0)
+			if contentLengthStr != "" {
+				size, err = strconv.ParseInt(contentLengthStr, 10, 64)
+				if err != nil {
+					log.Errorln("problem converting content-length to an int:", err)
+				}
+			}
+			headChan <- struct {
+				idx  int
+				size uint64
+				err  error
+			}{idx, uint64(size), nil}
+		}(idx, tUrl.String())
+	}
+	finished := make(map[int]bool)
+	for ctr := 0; ctr != len(attempts); ctr++ {
+		result := <-headChan
+		if result.err != nil {
+			if result.err != context.Canceled {
+				log.Debugf("Failure when doing a HEAD request against %s: %s", attempts[result.idx].Url.String(), result.err.Error())
+			}
+		} else {
+			finished[result.idx] = true
+			if result.idx == 0 {
+				cancel()
```

Comment:
Copy comment from hidden commit history:

> Trying to understand the logic here. Are we cancelling the reminding request if we get back from the first attempt? Wouldn't that cause the remaining attempts (no matter they succeeded or not) to have finished[idx] = false? Not sure if this will cause unwanted side effect